File: 2020/julia/src/Day2.jl
module Day2

include("Aoc.jl")
using .Aoc

passwordlist = Aoc.input_string(2)

regex = r"([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)\n"

struct Rule
    min::Int32
    max::Int32
    letter::Char
    password::String

    function Rule(match::RegexMatch)
        (min, max, letter, password) = match.captures
        new(parse(Int32, min), parse(Int32, max), letter[1], password)
    end
end

rules = [Rule(m) for m in eachmatch(regex, passwordlist)]

part1() = count(r -> r.min <= count(c -> c == r.letter, r.password) <= r.max, rules)
part2() = count(r -> (r.password[r.min] == r.letter || r.password[r.max] == r.letter) && r.password[r.min] != r.password[r.max], rules)

end

File: 2020/php/src/Aoc/Days/Day2.php
<?php


namespace Ppx17\Aoc2020\Aoc\Days;


use Illuminate\Support\Collection;

class Day2 extends AbstractDay
{
    private Collection $matches;

    public function setUp(): void
    {
        preg_match_all(
            "#([0-9]+)-([0-9]+) ([a-z]): ([a-z]+)#",
            $this->getInput(),
            $matches,
            PREG_SET_ORDER
        );

        $this->matches = collect($matches);
    }

    public function dayNumber(): int
    {
        return 2;
    }

    public function part1(): string
    {
        return $this->countWith(function($match, $min, $max, $letter, $password) {
            $occurrences = substr_count($password, $letter);
            return ($occurrences >= $min && $occurrences <= $max);
        });
    }

    public function part2(): string
    {
        return $this->countWith(function($match, $min, $max, $letter, $password) {
            return ($password[$min - 1] === $letter xor $password[$max - 1] === $letter);
        });
    }

    private function countWith(callable $callable) {
        return $this
            ->matches
            ->filter(fn($match) => $callable(...$match))
            ->count();
    }
}